### PR TITLE
📦 Declare mas as a Swift Package

### LINKED ISSUE #326: Allow running mas as portable or installing it
Thanks for creating and maintaining awesome tool ❤️

I wonder if there's a way to download and run a separate version `mas` and then deleting it without affecting existing installation of `mas`. I'd appreciate any suggestions on how to do that or feedback about "potential feature requests" part of my issue.

**Goal**

I'd like to be able download a specific version of the tool and use it in an isolated way. I'd possibly mess up existing installation of `mas` if I use `brew` or install it on the default location by running the installer.
 
**Potential feature requests**

-  Allow specification of installation path when I run `installer -pkg './mas.pkg' -target "/i-want-mas-to-be-here/"
- Distribution of portable version of mas without requiring any installer.

**Background/why**

I'll add some tweaks in  [privacy.sexy](https://github.com/undergroundwires/privacy.sexy) to remove preinstalled applications. And I'd like to distribute `mas` (download and run it on users' computers) to reinstall the applications (as a revert logic to removing them).

### PR DESCRIPTION
**Pull request by chris-araman**
- Removes dependence on `carthage` and `make`.
- Moves sources to `Sources` and `Tests` as is the default in new Swift packages.
  - If others would prefer these remain as they were, we can instead specify the paths explicitly in `Package.swift`.
- Declares Swift package dependencies in Xcode project and `Package.swift`.
- Removes the Xcode project and workspace. (See below.)
- `swift build` and `swift test` succeed.
- Updates scripts to use `swift build` and `swift test`. These continue to build universal (fat, arm64 and x86_64) binaries for macOS 10.11 and later. In the future, we could build for one architecture or the other, and have platform-specific bottles as many other formulae do.

To Do:
- [x] ~~Decide whether we want to remove the Xcode project and add `swift package generate-xcodeproj` to `script/bootstrap` instead. This would allow us to avoid maintaining multiple ways of building `mas`, and keeping duplicate dependency lists. The build scripts are still building the Xcode project instead of the Swift Package.~~
It turns out that `generate-xcodeproj` is deprecated and unnecessary. Xcode can open the `mas` directory directly because it's now a Swift Package.
- [x] ~~Decide if it's okay to link `MasKit` as a library into `mas` instead of deploying it as a framework. SPM does not yet support building frameworks, only libraries and executables.~~
I'm moving forward with the assumption that treating this as an SPM library is fine moving forward. If Apple decides in the future to add support for frameworks, we can add a corresponding product or target to `Package.swift`. The `mas` binary can now be relocated (#326). The Homebrew bottles are now much smaller.
- [x] ~~Decide if we want to keep our `Info.plist`. SPM does not have a supported way to process and inject these, though we could probably manage to write some `.linkerSettings(.unsafeFlags(...))` [hackery](https://gist.github.com/4np/f0e811bc0fdeb17186088c47d5bead5a) that would get the job done. It looks like we only inject copyright and version strings. The `version` command currently uses the injected version string. We could inject that in a different way.~~
I've added logic to `script/version` to generate `Sources/MasKit/Package.swift` with a string constant `Package.Version`. The `VersionCommand` now emits this constant instead of `CFBundleShortVersionString`. The copyright still exists in `LICENSE`, though not in the compiled binary. Should I add a `copyright` command or emit the copyright info from `help`?
- [x] ~~Resolve the GitHub Action test failures.~~
- [x] ~~Fix the `mas.pkg` build.~~

Fixes #326.

### CODE REVIEW

**Review by phatblat — APPROVED** [review]

**Comment by phatblat** [comment]
> Sorry I typed all this up last night and forgot to submit.

Yay! I've been wanting to do this for so long, but have not gotten around to it.

## To generate-xcodeproj or not to generate-xcodeproj?

I like the idea of generating the `xcodeproj` in the bootstrap script. I imagine that we'll ignore the file so that we don't check it in accidentally?

## MasKit framework

The only real reason MasKit exists is because the Xcode macOS CLI target doesn't support unit tests, or any embedded resources for that matter. I've come across one macOS GUI app that interfaced with the app store but they rolled their own hacks to the private frameworks; I'm not aware of any apps using MasKit.framework. I know that there are CLI dependencies on mas such as Chef, Ansible and ad-hoc scripts.

I believe we can change this packaging format to a library with little to no impact. But, since the separate framework was only created to facilitate testing, perhaps we no longer need it. Can we have the tests interface with any `@testable` internal API even if they are in an executable?

## Info.plist

The only use of `Info.plist` is that the version command fetches the value from `CFBundleShortVersionString`. If we can find a new home for the current version number (somewhere that's easy to remember) then we can lose the `Info.plist` file.

**Comment by chris-araman** [comment]
> `generate-xcodeproj`
No longer necessary. Xcode can open the `mas` directory. It's magic! 🪄✨

> MasKit framework
Cool. Yeah, the tests now run against the MasKit library.

> Info.plist
Great! They're gone.

I think I have everything working except for the `mas.pkg` build now. Working on that next. Feel free to accept and merge this when you're ready, or wait for `mas.pkg`. I can just submit that as a separate PR if need be.

**Comment by chris-araman** [comment]
The `mas.pkg` build works now. I've verified that installing it works as well. This PR is now ready to go as soon as you've accepted it.

**Comment by chris-araman** [comment]
> If we can find a new home for the current version number (somewhere that's easy to remember) then we can lose the Info.plist file.

The new home is in `git tag`. `script/version` will now fetch the latest git tag on the current branch, strip the `v` prefix, and generate `Sources/MasKit/Package.swift` using the tag as the `Package.Version` constant used by `VersionCommand`.

**Comment by rgoldberg** [comment]
Will a new version be released once this & #352 are merged?

Some apps are no being upgraded by `mas upgrade` in 1.8.1. I want to investigate & possibly fix these issues, but it doesn't make sense to do so until your recent changes have been merged into `master`, and preferably released, too.

**Comment by chris-araman** [comment]
@rgoldberg, I hope so. I have some other proposals (https://github.com/mas-cli/mas/pull/346#issuecomment-827123829, https://github.com/mas-cli/mas/issues/344#issuecomment-829374217, https://github.com/mas-cli/mas/issues/309#issuecomment-827966033) in the works, but none of them should prevent this PR and the previous ones from being released.

**Comment by rgoldberg** [comment]
Cool. Best to release everything that's almost ready ASAP, instead of delaying those things for new PRs.

**Comment by chris-araman** [comment]
@rgoldberg, you're a member of the mas-cli organization. Are you available to review this PR?

**Comment by phatblat** [comment]
@rgoldberg @chris-araman FYI the 3 of us are the only active members of the mas-cli org. I would like to get a [release](https://github.com/mas-cli/mas/releases/tag/untagged-42110dadf15c18842d90) out soon and think we're still looking at a patch release (1.8.2) with bug fixes and (many great) project improvements.

One piece that I haven't figured out yet is what to do about the hosting for the bottle files for our custom [homebrew-tap](https://github.com/mas-cli/homebrew-tap) since [Bintray has be sunsetted](https://jfrog.com/blog/into-the-sunset-bintray-jcenter-gocenter-and-chartcenter/). This shouldn't block the release.

**Comment by rgoldberg** [comment]
Unfortunately, I know almost nothing about Xcode, Swift, Apple APIs, etc. I'm a JVM developer. the `mas update` / `mas outdated` PR was the first time I coded in Swift or used Xcode. I know nothing about standard Apple project structures, SPMs, Frameworks, etc., so I don't have the requisite knowledge to review this. I'd happily review it, if I were qualified. I can, however, tinker around to investigate the few issues I've run into, and try to fix them, when it's more of straight ahead programming rather than knowing the ins & outs of the whole Apple ecosystem / platform. Sorry. Wish I could be more help.

**Comment by phatblat** [comment]
No worries, @rgoldberg! Any help is very appreciated! You are welcome to propose changes even if you're not sure if the code is right. @chris-araman and I can help with suggestions. Or, we work out the Swift and lower level stuff with your input on testing and reproducing issues.

**Comment by chris-araman** [comment]
> One piece that I haven't figured out yet is what to do about the hosting for the bottle files for our custom [homebrew-tap](https://github.com/mas-cli/homebrew-tap) since [Bintray has be sunsetted](https://jfrog.com/blog/into-the-sunset-bintray-jcenter-gocenter-and-chartcenter/). This shouldn't block the release.

It looks like Homebrew has migrated to GitHub Packages. Perhaps we should too?

**Comment by phatblat** [comment]
> It [looks like](https://github.com/Homebrew/discussions/discussions/691#discussioncomment-584299) Homebrew has migrated to GitHub Packages. Perhaps we should too?
> 
> I've opened [mas-cli/homebrew-tap#19](https://github.com/mas-cli/homebrew-tap/issues/19) for tracking.

👍🏻 My thoughts exactly 😉

**Comment by phatblat** [comment]
@chris-araman I'm re-reviewing this PR so we can merge today.

File: Sources/MasKit/Formatters/AppListFormatter.swift [hunk 1]
```diff
@@ -19,18 +19,14 @@ struct AppListFormatter {
...
-            ?? nameColumnMinWidth
+        let maxLength = products.map(\.appNameOrBundleIdentifier.count).max() ?? nameColumnMinWidth
 
         var output: String = ""
 
         for product in products {
             let appId = product.itemIdentifier.stringValue
                 .padding(toLength: idColumnMinWidth, withPad: " ", startingAt: 0)
-            let appName = product.appNameOrBbundleIdentifier.padding(toLength: maxLength, withPad: " ", startingAt: 0)
+            let appName = product.appNameOrBundleIdentifier.padding(toLength: maxLength, withPad: " ", startingAt: 0)
```
Comment (phatblat) [inline Sources/MasKit/Formatters/AppListFormatter.swift:29]: Oops! Thanks for fixing this typo.

File: Sources/MasKit/Formatters/Utilities.swift [hunk 2]
```diff
@@ -13,6 +13,22 @@ import Foundation
...
+
+// Override global print for testability.
+// See MasKitTests/OutputListener.swift.
+func print(_ items: Any..., separator: String = " ", terminator: String = "\n") {
+    if let observer = printObserver {
+        let output = items
+            .map { "\($0)" }
+            .joined(separator: separator)
+            .appending(terminator)
+        observer(output)
```
Comment (phatblat) [inline Sources/MasKit/Formatters/Utilities.swift:26]: Fancy!

File: Sources/MasKit/Models/SoftwareProduct.swift [hunk 3]
```diff
@@ -29,8 +29,8 @@ extension SoftwareProduct {
     }
 
     /// Returns bundleIdentifier if appName is empty string.
-    var appNameOrBbundleIdentifier: String {
-        appName == "" ? bundleIdentifier : appName
+    var appNameOrBundleIdentifier: String {
```
Comment (phatblat) [inline Sources/MasKit/Models/SoftwareProduct.swift:32]: 👍🏻

File: Sources/MasKit/Formatters/Utilities.swift (hunk 2 above)
Comment (chris-araman) [reply Sources/MasKit/Formatters/Utilities.swift:26]: Or an ugly hack, depending on your point of view. 😉

File: script/release [hunk 4]
```diff
@@ -1,86 +0,0 @@
-#!/bin/bash -e
-#
-# script/release
-# mas
-#
-# Updates app version in Info.plist file and tags release commit.
-# Not currently in use.
```
Comment (phatblat) [inline script/release:7]: 👍🏻 I'll document the steps I go through to release. I'd love to have a github action on release which won't need tagging as the GitHub release creates a tag.

Comment (chris-araman) [reply script/release:7]: Sounds good. I was realizing that I don't know the process as I was working on this PR.

File: Sources/MasKit/Commands/Version.swift [hunk 5]
```diff
@@ -18,10 +18,7 @@ public struct VersionCommand: CommandProtocol {
 
     /// Runs the command.
     public func run(_: Options) -> Result<Void, MASError> {
-        let plist = Bundle.main.infoDictionary
-        if let versionString = plist?["CFBundleShortVersionString"] {
-            print(versionString)
-        }
+        print(Package.Version)
```
Comment (phatblat) [inline Sources/MasKit/Commands/Version.swift:21]: I'm seeing extra characters in the `version` command output.

```
↪ .build/x86_64-apple-macosx/debug/mas version
["1.8.1"]
```

Any ideas where the quotes and brackets are coming from?

**Review by phatblat — CHANGES_REQUESTED** [review]
Fantastic work! I love seeing deletes since they are removing cruft.

I found one issue with the `version` command that I'd like you to look into.

File: Sources/MasKit/Commands/Version.swift (hunk 5 above)
Comment (chris-araman) [reply Sources/MasKit/Commands/Version.swift:21]: 🙈 Fixed!

**Review by phatblat — APPROVED** [review]

**Pull request merged**